Model/Indexer/Invalidate.php

```diff
@@ -0,0 +1,158 @@
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * @author Nosto Solutions Ltd <contact@nosto.com>
+ * @copyright 2019 Nosto Solutions Ltd
+ * @license http://opensource.org/licenses/BSD-3-Clause BSD 3-Clause
+ *
+ */
+
+namespace Nosto\Tagging\Model\Indexer;
+
+use Exception;
+use Magento\Framework\Indexer\ActionInterface as IndexerActionInterface;
+use Magento\Framework\Mview\ActionInterface as MviewActionInterface;
+use Magento\Store\Model\Store;
+use Nosto\Tagging\Helper\Account as NostoHelperAccount;
+use Nosto\Tagging\Model\ResourceModel\Magento\Product\Collection as ProductCollection;
+use Nosto\Tagging\Model\ResourceModel\Magento\Product\CollectionFactory as ProductCollectionFactory;
+use Nosto\Tagging\Model\Service\Index as NostoServiceIndex;
+use Nosto\Tagging\Util\Indexer as IndexerUtil;
+
+/**
+ * Class Invalidate
+ * This class is responsible for listening to product changes
+ * and setting the `is_dirty` value in `nosto_product_index` table
+ * @package Nosto\Tagging\Model\Indexer
+ */
+class Invalidate implements IndexerActionInterface, MviewActionInterface
+{
+    const INDEXER_ID = 'nosto_index_product_invalidate';
+
+    /** @var NostoHelperAccount */
+    private $nostoHelperAccount;
+
+    /** @var NostoServiceIndex */
+    private $nostoServiceIndex;
+
+    /** @var ProductCollectionFactory */
+    private $productCollectionFactory;
+
+    /**
+     * Dirty constructor.
+     * @param NostoHelperAccount $nostoHelperAccount
+     * @param NostoServiceIndex $nostoServiceIndex
+     * @param ProductCollectionFactory $productCollectionFactory
+     */
+    public function __construct(
+        NostoHelperAccount $nostoHelperAccount,
+        NostoServiceIndex $nostoServiceIndex,
+        ProductCollectionFactory $productCollectionFactory
+    ) {
+        $this->nostoHelperAccount = $nostoHelperAccount;
+        $this->nostoServiceIndex = $nostoServiceIndex;
+        $this->productCollectionFactory = $productCollectionFactory;
+    }
+
+    /**
+     * @param int[] $ids
+     * @throws Exception
+     */
+    public function execute($ids)
+    {
+        if (!empty($ids)) {
+            $ids = array_unique($ids);
+            $idsSize = count($ids);
+            $storesWithNosto = $this->nostoHelperAccount->getStoresWithNosto();
+            foreach ($storesWithNosto as $store) {
+                $productCollection = $this->getCollection($store, $ids);
+                $this->nostoServiceIndex->invalidateOrCreate($productCollection, $store);
+                $collectionSize = $productCollection->getSize();
+
+                if ($idsSize > $collectionSize) {
```

Comment:
What is this?

Reply:
The logic behind this is that if the Magento passes product ids that have been deleted to our indexer we won't find those anymore from the db. When products are not in db anymore we consider those as deleted.